egs/librispeech/ASR/finetune_hubert_transducer/hubert_encoder.py

```diff
@@ -0,0 +1,260 @@
+                )
+            )
+            model.mask_channel_prob = mask_channel_prob
+        if mask_prob != model.mask_prob:
+            logging.warning(
+                "Overwriting mask prob to {}. Original ckpt to {}".format(
+                    mask_prob, model.mask_prob
+                )
+            )
+            model.mask_prob = mask_prob
+
+        # CNN feature extractor is frozen during finetuning!
+        self.encoders.feature_grad_mult = 0.0
+
+        self.freeze_finetune_updates = freeze_finetune_updates
+        self.register_buffer("num_updates", torch.LongTensor([0]))
+
+        if subsample_output:
+            self.subsample = get_subsample_module(
+                subsample_mode,
+                input_size=self.hubert_output_size,
+                output_size=output_size,
+            )
+            self.subsample_mode = subsample_mode
+            logging.info("Subsample output!}")
+            logging.info(self.subsample)
+        else:
+            self.subsample = None
+            self.subsample_mode = None
+            logging.info("Do not subsample output")
+
+        self.training = training
+        delattr(self.encoders, "final_proj")
+        delattr(self.encoders, "label_embs_concat")
+
+    @classmethod
+    def add_arguments(cls, parser):
+        parser.add_argument(
+            "--hubert-model-dir",
+            type=str,
+            help="Path to the pretrained Hubert model",
+        )
+
+        parser.add_argument(
+            "--hubert-freeze-finetune-updates",
+            type=int,
+            default=0,
+            help="The number of updates during which the transformer blocks \
+                in hubert are frozen.",
+        )
+
+        parser.add_argument(
+            "--hubert-mask-prob",
+            type=float,
+            default=0.65,
+            help="Mask probability",
+        )
+
+        parser.add_argument(
+            "--hubert-mask-channel-prob",
+            type=float,
+            default=0.5,
+            help="Mask channel probability",
+        )
+
+        parser.add_argument(
+            "--hubert-mask-channel-length",
+            type=int,
+            default=64,
+            help="Mask channel length",
+        )
+
+        parser.add_argument(
+            "--hubert-subsample-output",
+            type=str2bool,
+            default=False,
+            help="Whether subsample the hubert output to reduce frame rate",
+        )
+
+        parser.add_argument(
+            "--hubert-subsample-mode",
+            type=str,
+            default="concat_tanh",
+            choices=["concat", "concat_relu", "concat_tanh", "avgpooling"],
+        )
+
+    def normalise_input(
+        self, x: torch.Tensor, x_lens: torch.Tensor
+    ) -> torch.Tensor:
+        """Normalize a batch of audio samples to zero mean unit variance
+
+        Args:
+            x (torch.Tensor): a batch of audio samples
```

Comment:
Could you add more details about x and x_lens? E.g., the shape of `x` and `x_lens`; is `x` right padded with 0s for some uttterances?

Reply:
Done